[CBRD-20561] value of Num_plan_cache_entries is always 0

asturzu opened the pull request:
http://jira.cubrid.org/browse/CBRD-20561

The bug was that the value of Num_plan_cache_entries is always 0. The first problem
was that in xasl_cache.c the counter wasn't always incremented.
The second problem was that the approach that I first used for peek statistics was wrong because the global variables that I used should have been sent from the server to the client.
In order to fix this I used the global array for statistics and at request I copied for the special peek statistics the values from the global statistics array to the local thread statistics array.

eseokoh commented:
The change on copying stats to global looks correct. 
I'd like to confirm the change on `xcache_Entry_count` again. If I'm correct, we had a regression on counting the number of entries and added `else if (inserted)` block to fix the regression.

cubridci commented:
Build comment file: 
--------------
#### File: src/base/perf_monitor.c (3 incorrect indentation(s) found)
```
  4488   stats[pstat_Metadata[PSTAT_PC_NUM_CACHE_ENTRIES].start_offset] = get_xcache_entry_count ();
  4489   stats[pstat_Metadata[PSTAT_HF_NUM_STATS_ENTRIES].start_offset] = get_num_stats_entries ();
  4490   stats[pstat_Metadata[PSTAT_QM_NUM_HOLDABLE_CURSORS].start_offset] = get_number_of_holdable_cursors ();
```
#### File: src/query/xasl_cache.c (3 incorrect indentation(s) found)
```
  1128 
  2125 int
  2126 get_xcache_entry_count ()
```
#### File: src/session/session.c (3 incorrect indentation(s) found)
```
  2584 
  2992 int
  2993 get_number_of_holdable_cursors ()
```
#### File: src/storage/heap_file.c (5 incorrect indentation(s) found)
```
  1147 
  1180 
 24884  */
 24885 int
 24886 get_num_stats_entries ()
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

asturzu commented:
Yes,

I misunderstood that part of the code. I reverted the change. Also, I
modified to get the peek statistics values exactly
at the request. In this way I eliminated the set calls made for the global
statistics array.

        Best regards,
           Toni

On Thu, Oct 20, 2016 at 11:21 AM, ESeok Oh <notifications@github.com> wrote:

> The change on copying stats to global looks correct.
> I'd like to confirm the change on xcache_Entry_count again. If I'm
> correct, we had a regression on counting the number of entries and added else
> if (inserted) block to fix the regression.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CUBRID/cubrid/pull/287#issuecomment-255040481>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQVQwmh5kDCsX4BGWtXD6xJzuqVUdvE4ks5q1ySPgaJpZM4KXtqd>
> .
>

ribram commented on src/storage/heap_file.c:
```diff
@@ -24880,3 +24876,14 @@ heap_rv_undo_ovf_update (THREAD_ENTRY * thread_p, LOG_RCV * rcv)
 }
+
+/*
+ * get_num_stats_entries - Returns the number of num_stats_entries 
```
rename as heap_get_best_space_num_stats_entries

ribram commented on src/base/perf_monitor.c:
```diff
@@ -1835,6 +1830,7 @@ perfmon_server_get_stats (THREAD_ENTRY * thread_p)
       return NULL;
     }
 
+  copy_peek_stats (pstat_Global.tran_stats[tran_index]);
```
rename as perfmon_get_peek_stats or something similar.

eseokoh commented on src/query/xasl_cache.c:
```diff
@@ -2118,3 +2115,15 @@ xcache_check_recompilation_threshold (THREAD_ENTRY * thread_p, XASL_CACHE_ENTRY
+ * return : the number of xasl cache entries
+ */
+int
+get_xcache_entry_count ()
```
and also fill the parameters with `void`:

```c
int
xcache_get_entry_count (void)
```

eseokoh commented on src/query/xasl_cache.h:
```diff
@@ -114,4 +114,5 @@ extern void xcache_dump (THREAD_ENTRY * thread_p, FILE * fp);
 extern bool xcache_can_entry_cache_list (XASL_CACHE_ENTRY * xcache_entry);
 
 extern void xcache_retire_clone (THREAD_ENTRY * thread_p, XASL_CACHE_ENTRY * xcache_entry, XASL_CLONE * xclone);
+extern int get_xcache_entry_count ();
```
```
extern int xcache_get_entry_count (void);
```

eseokoh commented on src/session/session.c:
```diff
@@ -2985,3 +2982,15 @@ session_state_decrease_ref_count (THREAD_ENTRY * thread_p, SESSION_STATE * state
+ * 
+ */
+int
+get_number_of_holdable_cursors ()
```
Please also fill `void`.

eseokoh commented on src/session/session.h:
```diff
@@ -71,4 +71,5 @@ extern int session_get_trace_stats (THREAD_ENTRY * thread_p, DB_VALUE * result);
 extern int session_set_trace_stats (THREAD_ENTRY * thread_p, char *scan_stats, int format);
 extern int session_clear_trace_stats (THREAD_ENTRY * thread_p);
 extern TZ_REGION *session_get_session_tz_region (THREAD_ENTRY * thread_p);
+extern int get_number_of_holdable_cursors ();
```
Please also fill `void`.

eseokoh commented on src/storage/heap_file.c:
```diff
@@ -24880,3 +24876,14 @@ heap_rv_undo_ovf_update (THREAD_ENTRY * thread_p, LOG_RCV * rcv)
+ *
+ */
+int
+get_num_stats_entries ()
```
please also fill `void`.

eseokoh commented on src/storage/heap_file.h:
```diff
@@ -669,5 +669,6 @@ extern SCAN_CODE heap_get_visible_version_internal (THREAD_ENTRY * thread_p, HEA
 extern SCAN_CODE heap_get_class_record (THREAD_ENTRY * thread_p, const OID * class_oid, RECDES * recdes_p,
 					HEAP_SCANCACHE * scan_cache, int ispeeking);
 extern int heap_rv_undo_ovf_update (THREAD_ENTRY * thread_p, LOG_RCV * rcv);
+extern int get_num_stats_entries ();
```
please also fill `void`.
